source/extensions/geoip_providers/maxmind/geoip_provider.cc

```diff
@@ -0,0 +1,279 @@
+      if (status == MMDB_SUCCESS && entry_data_list) {
+        populateGeoLookupResult(mmdb_lookup_result, lookup_result, config_->asnHeader().value(),
+                                MMDB_ASN_LOOKUP_ARGS[0]);
+        // todo remove this in final version, this is only for testing
+        MMDB_dump_entry_data_list(stdout, entry_data_list, 2);
+        MMDB_free_entry_data_list(entry_data_list);
+        if (lookup_result.size() > n_prev_hits) {
+          config_->incHit("isp_db");
+        }
+      } else {
+        config_->incLookupError("isp_db");
+      }
+    }
+    config_->incTotal("isp_db");
+  }
+}
+
+// todo: document En as default and only language used for lookups. Document that region is selcted
+// as least specific subdivision
+void GeoipProvider::lookupInAnonDb(
+    const Network::Address::InstanceConstSharedPtr& remote_address,
+    absl::flat_hash_map<std::string, std::string>& lookup_result) const {
+  if (config_->isLookupEnabledForHeader(config_->anonHeader()) || config_->anonVpnHeader()) {
+    ASSERT(anon_db_, "Maxmind city database is not initialised for performing lookups");
+    int mmdb_error;
+    const uint32_t n_prev_hits = lookup_result.size();
+    MMDB_lookup_result_s mmdb_lookup_result = MMDB_lookup_sockaddr(
+        anon_db_.get(), reinterpret_cast<const sockaddr*>(remote_address->sockAddr()), &mmdb_error);
+    if (!mmdb_error) {
+      MMDB_entry_data_list_s* entry_data_list;
+      int status = MMDB_get_entry_data_list(&mmdb_lookup_result.entry, &entry_data_list);
+      if (status == MMDB_SUCCESS) {
+        if (config_->isLookupEnabledForHeader(config_->anonHeader())) {
+          populateGeoLookupResult(mmdb_lookup_result, lookup_result, config_->anonHeader().value(),
+                                  MMDB_ANON_LOOKUP_ARGS[0]);
+        }
+        if (config_->isLookupEnabledForHeader(config_->anonVpnHeader())) {
+          populateGeoLookupResult(mmdb_lookup_result, lookup_result,
+                                  config_->anonVpnHeader().value(), MMDB_ANON_LOOKUP_ARGS[1]);
+        }
+        if (config_->isLookupEnabledForHeader(config_->anonHostingHeader())) {
+          populateGeoLookupResult(mmdb_lookup_result, lookup_result,
+                                  config_->anonHostingHeader().value(), MMDB_ANON_LOOKUP_ARGS[2]);
+        }
+        if (config_->isLookupEnabledForHeader(config_->anonTorHeader())) {
+          populateGeoLookupResult(mmdb_lookup_result, lookup_result,
+                                  config_->anonTorHeader().value(), MMDB_ANON_LOOKUP_ARGS[3]);
+        }
+        if (config_->isLookupEnabledForHeader(config_->anonProxyHeader())) {
+          populateGeoLookupResult(mmdb_lookup_result, lookup_result,
+                                  config_->anonProxyHeader().value(), MMDB_ANON_LOOKUP_ARGS[4]);
+        }
+        // todo remove this in final version, this is only for testing
+        MMDB_dump_entry_data_list(stdout, entry_data_list, 2);
```

Comment:
Seems like there's a lot of these `todo` notes in here, please either address them or format them as `TODO(nezdolik):` if they're actually intended for later.

Reply:
True, going through all of them now, also need to update user documentation and would be good to add a configuration example.